File: app/src/main/java/org/lineageos/recorder/DialogActivity.java
/*
 * Copyright (C) 2017-2021 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.lineageos.recorder;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatEditText;
import androidx.appcompat.widget.SwitchCompat;

import org.lineageos.recorder.utils.PermissionManager;
import org.lineageos.recorder.utils.PreferencesManager;

public class DialogActivity extends AppCompatActivity {
    public static final String EXTRA_IS_RECORDING = "is_recording";

    private PermissionManager mPermissionManager;
    private PreferencesManager mPreferences;
    private SwitchCompat mLocationSwitch;
    private SwitchCompat mCircularSwitch;

    @Override
    protected void onCreate(@Nullable Bundle savedInstance) {
        super.onCreate(savedInstance);

        mPermissionManager = new PermissionManager(this);
        mPreferences = new PreferencesManager(this);

        setFinishOnTouchOutside(true);

        final AlertDialog dialog = new AlertDialog.Builder(this)
                .setTitle(R.string.settings_title)
                .setView(R.layout.dialog_content_settings)
                .setOnDismissListener(dialogInterface -> finish())
                .show();

        final boolean isRecording = getIntent().getBooleanExtra(EXTRA_IS_RECORDING, false);

        mLocationSwitch = dialog.findViewById(
                R.id.dialog_content_settings_location_switch);
        if (mLocationSwitch != null) {
            setupLocationSwitch(mLocationSwitch, isRecording);
        }
        final SwitchCompat highQualitySwitch = dialog.findViewById(
                R.id.dialog_content_settings_high_quality_switch);
        if (highQualitySwitch != null) {
            setupHighQualitySwitch(highQualitySwitch, isRecording);
        }
        mCircularSwitch = dialog.findViewById(
                R.id.dialog_content_settings_circular_recording_switch);
        if (mCircularSwitch != null) {
            setupCircularSwitch(mCircularSwitch, isRecording);
        }
        final AppCompatEditText circularPeriodInput = dialog.findViewById(
                R.id.dialog_content_settings_circular_recording_period_input);
        if (circularPeriodInput != null) {
            setupCircularPeriodInput(circularPeriodInput, isRecording);
        }
        final AppCompatEditText circularNumberInput = dialog.findViewById(
                R.id.dialog_content_settings_circular_recording_number_input);
        if (circularNumberInput != null) {
            setupCircularNumberInput(circularNumberInput, isRecording);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] results) {
        super.onRequestPermissionsResult(requestCode, permissions, results);
        if (requestCode == PermissionManager.REQUEST_CODE) {
            if (mPermissionManager.hasLocationPermission()) {
                toggleAfterPermissionRequest();
            } else {
                mPermissionManager.onLocationPermissionDenied();
                mLocationSwitch.setChecked(false);
            }
        }
    }

    @Override
    public void onBackPressed() {
        finish();
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(0, android.R.anim.fade_out);
    }

    private void setupLocationSwitch(@NonNull SwitchCompat locationSwitch,
                                     boolean isRecording) {
        final boolean tagWithLocation;
        if (mPreferences.getTagWithLocation()) {
            if (mPermissionManager.hasLocationPermission()) {
                tagWithLocation = true;
            } else {
                // Permission revoked -> disabled feature
                mPreferences.setTagWithLocation(false);
                tagWithLocation = false;
            }
        } else {
            tagWithLocation = false;
        }

        locationSwitch.setChecked(tagWithLocation);

        if (isRecording) {
            locationSwitch.setEnabled(false);
        } else {
            locationSwitch.setOnCheckedChangeListener((button, isChecked) -> {
                if (isChecked) {
                    if (mPermissionManager.hasLocationPermission()) {
                        mPreferences.setTagWithLocation(true);
                    } else {
                        mPermissionManager.requestLocationPermission();
                    }
                } else {
                    mPreferences.setTagWithLocation(false);
                }
            });
        }
    }

    private void setupHighQualitySwitch(@NonNull SwitchCompat highQualitySwitch,
                                        boolean isRecording) {
        final boolean highQuality = mPreferences.getRecordInHighQuality();
        highQualitySwitch.setChecked(highQuality);

        if (isRecording) {
            highQualitySwitch.setEnabled(false);
        } else {
            highQualitySwitch.setOnCheckedChangeListener((button, isChecked) ->
                    mPreferences.setRecordingHighQuality(isChecked));
        }
    }

    // basically a copy of setupLocationSwitch
    private void setupCircularSwitch(@NonNull SwitchCompat circularSwitch,
                                     boolean isRecording) {
        final boolean circularRecording;
        if (mPreferences.getCircularRecording()) {
            if (mPermissionManager.hasBatteryPermission()) {
                circularRecording = true;
            } else {
                // Permission revoked -> disabled feature
                mPreferences.setCircularRecording(false);
                circularRecording = false;
            }
        } else {
            circularRecording = false;
        }

        circularSwitch.setChecked(circularRecording);

        if (isRecording) {
            circularSwitch.setEnabled(false);
        } else {
            circularSwitch.setOnCheckedChangeListener((button, isChecked) -> {
                if (isChecked) {
                    if (mPermissionManager.hasBatteryPermission()) {
                        mPreferences.setCircularRecording(true);
                    } else {
                        mPermissionManager.requestBatteryPermission();
                    }
                } else {
                    mPreferences.setCircularRecording(false);
                }
            });
        }
    }

    private void setupCircularPeriodInput(@NonNull AppCompatEditText input, boolean isRecording) {
        final long val = mPreferences.getCircularRecordingPeriod();
        input.setText(new java.text.DecimalFormat("0.#").format(((double)val) / 60.0));

        if (isRecording) {
            input.setEnabled(false);
        } else {
            input.addTextChangedListener(new TextWatcher() {
                @Override
                public void afterTextChanged(Editable s) {
                    final String str = s.toString();
                    if (str.length() > 0) {
                        mPreferences.setCircularRecordingPeriod((long)(Double.parseDouble(str) * 60.0));
                    }
                }
                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {}
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            });
        }
    }

    private void setupCircularNumberInput(@NonNull AppCompatEditText input, boolean isRecording) {
        final int val = mPreferences.getCircularRecordingNumber();
        input.setText(Integer.toString(val));

        if (isRecording) {
            input.setEnabled(false);
        } else {
            input.addTextChangedListener(new TextWatcher() {
                @Override
                public void afterTextChanged(Editable s) {
                    final String str = s.toString();
                    if (str.length() > 0) {
                        mPreferences.setCircularRecordingNumber(Integer.parseInt(str));
                    }
                }
                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {}
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            });
        }
    }

    private void toggleAfterPermissionRequest() {
        mLocationSwitch.setChecked(true);
        mPreferences.setTagWithLocation(true);
    }
}

File: app/src/main/java/org/lineageos/recorder/DialogActivity.kt
/*
 * SPDX-FileCopyrightText: 2017-2024 The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.recorder

import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.widget.CompoundButton
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatEditText
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.materialswitch.MaterialSwitch
import org.lineageos.recorder.utils.PermissionManager
import org.lineageos.recorder.utils.PreferencesManager

class DialogActivity : AppCompatActivity() {
    // Views
    private lateinit var highQualitySwitch: MaterialSwitch
    private lateinit var locationSwitch: MaterialSwitch
    private lateinit var circularSwitch: MaterialSwitch

    private val permissionManager: PermissionManager by lazy { PermissionManager(this) }

    private val preferences by lazy { PreferencesManager(this) }

    override fun onCreate(savedInstance: Bundle?) {
        super.onCreate(savedInstance)

        setFinishOnTouchOutside(true)

        val dialog = MaterialAlertDialogBuilder(this)
            .setTitle(R.string.settings_title)
            .setView(R.layout.dialog_content_settings)
            .setOnDismissListener { finish() }
            .show()

        val isRecording = intent.getBooleanExtra(EXTRA_IS_RECORDING, false)
        locationSwitch = dialog.findViewById(R.id.locationSwitch)!!

        setupLocationSwitch(locationSwitch, isRecording)

        highQualitySwitch = dialog.findViewById(R.id.highQualitySwitch)!!
        setupHighQualitySwitch(highQualitySwitch, isRecording)

        circularSwitch = dialog.findViewById(R.id.circularSwitch)!!
        setupCircularSwitch(circularSwitch, isRecording)

        val circularPeriodInput: AppCompatEditText? = dialog.findViewById(
                R.id.dialog_content_settings_circular_recording_period_input)
        if (circularPeriodInput != null) {
            setupCircularPeriodInput(circularPeriodInput, isRecording)
        }
        val circularNumberInput: AppCompatEditText? = dialog.findViewById(
                R.id.dialog_content_settings_circular_recording_number_input)
        if (circularNumberInput != null) {
            setupCircularNumberInput(circularNumberInput, isRecording)
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>,
        results: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, results)
        if (requestCode == PermissionManager.REQUEST_CODE) {
            if (permissionManager.hasLocationPermission()) {
                toggleAfterPermissionRequest()
            } else {
                permissionManager.onLocationPermissionDenied()
                locationSwitch.isChecked = false
            }
        }
    }

    @Deprecated("Deprecated in Java")
    override fun onBackPressed() {
        finish()
    }

    override fun finish() {
        super.finish()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            overrideActivityTransition(
                OVERRIDE_TRANSITION_CLOSE, 0, android.R.anim.fade_out, Color.TRANSPARENT
            )
        } else {
            @Suppress("deprecation")
            overridePendingTransition(0, android.R.anim.fade_out)
        }
    }

    private fun setupLocationSwitch(
        locationSwitch: MaterialSwitch,
        isRecording: Boolean
    ) {
        val tagWithLocation = if (preferences.tagWithLocation) {
            if (permissionManager.hasLocationPermission()) {
                true
            } else {
                // Permission revoked -> disabled feature
                preferences.tagWithLocation = false
                false
            }
        } else {
            false
        }
        locationSwitch.isChecked = tagWithLocation
        if (isRecording) {
            locationSwitch.isEnabled = false
        } else {
            locationSwitch.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
                if (isChecked) {
                    if (permissionManager.hasLocationPermission()) {
                        preferences.tagWithLocation = true
                    } else {
                        permissionManager.requestLocationPermission()
                    }
                } else {
                    preferences.tagWithLocation = false
                }
            }
        }
    }

    private fun setupCircularSwitch(
        circularSwitch: MaterialSwitch,
        isRecording: Boolean
    ) {
        val circularRecording = if (preferences.circularRecording) {
            if (permissionManager.hasBatteryPermission()) {
                true
            } else {
                // Permission revoked -> disabled feature
                preferences.circularRecording = false
                false
            }
        } else {
            false
        }
        circularSwitch.isChecked = circularRecording
        if (isRecording) {
            circularSwitch.isEnabled = false
        } else {
            circularSwitch.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
                if (isChecked) {
                    if (permissionManager.hasBatteryPermission()) {
                        preferences.circularRecording = true
                    } else {
                        permissionManager.requestBatteryPermission()
                    }
                } else {
                    preferences.circularRecording = false
                }
            }
        }
    }

    private fun setupCircularPeriodInput(input: AppCompatEditText, isRecording: Boolean) {
        val v: Long = preferences.circularRecordingPeriod
        input.setText(java.text.DecimalFormat("0.#").format(v.toDouble() / 60.0))

        if (isRecording) {
            input.setEnabled(false)
        } else {
            input.addTextChangedListener(object : TextWatcher {
                override fun afterTextChanged(s: Editable) {
                    val str: String = s.toString()
                    if (str.length > 0) {
                        preferences.circularRecordingPeriod = (str.toDouble() * 60.0).toLong()
                    }
                }
                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            })
        }
    }

    private fun setupCircularNumberInput(input: AppCompatEditText, isRecording: Boolean) {
        val v: Int = preferences.circularRecordingNumber
        input.setText(Integer.toString(v))

        if (isRecording) {
            input.setEnabled(false)
        } else {
            input.addTextChangedListener(object : TextWatcher {
                override fun afterTextChanged(s: Editable) {
                    val str: String = s.toString()
                    if (str.length > 0) {
                        preferences.circularRecordingNumber = str.toInt()
                    }
                }
                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            })
        }
    }

    private fun setupHighQualitySwitch(
        highQualitySwitch: MaterialSwitch,
        isRecording: Boolean
    ) {
        val highQuality = preferences.recordInHighQuality
        highQualitySwitch.isChecked = highQuality
        if (isRecording) {
            highQualitySwitch.isEnabled = false
        } else {
            highQualitySwitch.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
                preferences.recordInHighQuality = isChecked
            }
        }
    }

    private fun toggleAfterPermissionRequest() {
        locationSwitch.isChecked = true
        preferences.tagWithLocation = true
    }

    companion object {
        const val EXTRA_IS_RECORDING = "is_recording"
    }
}
